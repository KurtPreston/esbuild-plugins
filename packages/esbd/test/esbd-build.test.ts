import { node } from 'execa';
import fs from 'fs';
import path from 'path';

import type { EsbdConfig } from '../lib';

import type { BuildWithHTMLOutput } from './config/serializer';

interface BuildWithHTMLOptions {
  config: Omit<EsbdConfig, 'absWorkingDir' | 'outdir'>;
  files: { [file: string]: string };
}

async function build(options: BuildWithHTMLOptions): Promise<BuildWithHTMLOutput> {
  await fs.promises.mkdir(path.join(__dirname, 'tests'), { recursive: true });
  const absWorkingDir = await fs.promises.mkdtemp(path.join(__dirname, 'tests', 'test-'));
  const absOutDir = path.join(absWorkingDir, 'out');

  await fs.promises.mkdir(absOutDir, { recursive: true });

  const config: EsbdConfig = {
    ...options.config,
    sourcemap: false,
    absWorkingDir,
    logLevel: 'warning',
    outdir: './out',
  };

  const bundleFile = path.join(absWorkingDir, 'bundle.js');
  const writeBundle = fs.promises.writeFile(
    bundleFile,
    `require('../../../lib').bundle(${JSON.stringify(config)});`,
  );

  const writeFiles = Object.entries(options.files).map(async ([file, content]) => {
    const absFilePath = path.join(absWorkingDir, file);
    await fs.promises.mkdir(path.dirname(absFilePath), { recursive: true });
    await fs.promises.writeFile(absFilePath, content, { encoding: 'utf-8' });
  });

  await Promise.all([...writeFiles, writeBundle]);

  const proc = node(bundleFile, ['build'], {
    encoding: 'utf8',
    reject: false,
    cwd: absWorkingDir,
    env: { ...process.env, NO_COLOR: '1' },
  });

  const { stderr, stdout } = await proc;

  return { outdir: absOutDir, stdout, stderr };
}

describe('build command', () => {
  afterAll(async () => {
    await fs.promises.rm(path.join(__dirname, 'tests'), { recursive: true });
  });

  it('builds a simple entry point', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from entry point', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx', style: 'styles/entry.css' },
          format: 'esm',
        },
        files: {
          'styles/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'styles/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from JS', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'src/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            import './entry.css';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('supports automatic react runtime', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          jsxRuntime: 'automatic',
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });
});
