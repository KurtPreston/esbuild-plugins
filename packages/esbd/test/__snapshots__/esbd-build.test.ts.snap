// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build command builds a simple entry point 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
"use strict";

// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

`;

exports[`build command includes referenced CSS from JS 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.css
---------------------------------
/* src/app.css */
.app {
  background: green;
}

/* src/entry.css */
body {
  background: red;
}

---------------------------------
entry.js
---------------------------------
"use strict";

// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

`;

exports[`build command includes referenced CSS from entry point 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry, style with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
"use strict";

// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
style.css
---------------------------------
/* styles/app.css */
.app {
  background: green;
}

/* styles/entry.css */
body {
  background: red;
}

`;

exports[`build command supports automatic react runtime 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
"use strict";
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) =>
  function __init() {
    return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])((fn = 0))), res;
  };
var __commonJS = (cb, mod) =>
  function __require() {
    return (
      mod ||
        (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod),
      mod.exports
    );
  };

// src/app.tsx
import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";
function App() {
  return /* @__PURE__ */ _jsxDEV(
    "div",
    {
      children: "Hello world",
    },
    void 0,
    false,
    {
      fileName: "<CWD>/src/app.tsx",
      lineNumber: 3,
      columnNumber: 22,
    },
    this
  );
}
var init_app = __esm({
  "src/app.tsx"() {
    "use strict";
  },
});

// src/entry.tsx
import { jsxDEV as _jsxDEV2 } from "react/jsx-dev-runtime";
import ReactDOM from "react-dom";
var require_entry = __commonJS({
  "src/entry.tsx"(exports) {
    init_app();
    ReactDOM.render(
      /* @__PURE__ */ _jsxDEV2(
        App,
        {},
        void 0,
        false,
        {
          fileName: "<CWD>/src/entry.tsx",
          lineNumber: 4,
          columnNumber: 29,
        },
        exports
      ),
      document.getElementById("root")
    );
  },
});
export default require_entry();

`;
